### PR TITLE
Add Limit to Tranasctional Layers

### PR DESCRIPTION
**Pull request by shawntabrizi**
Breaking down: https://github.com/paritytech/substrate/pull/10744

This PR simply adds a limit to the number of tranactional layers allowed within FRAME.

It sets the limit to 255.

It requires that all uses of `with_transactional` can return a `Result` where the error can be mapped to `DispatchError`.

We also provide a `with_transactional_unchecked` which is has the same behavior as before, and can be used (hopefully only temporarily), where it is not easy to update to the new required signature.

polkadot companion: https://github.com/paritytech/polkadot/pull/5232

### CODE REVIEW

File: frame/contracts/src/exec.rs [hunk 1]
```diff
@@ -766,14 +766,27 @@ where
...
+					_ => TransactionOutcome::Rollback(Ok((false, output))),
+				}
+			});
+
+		let (success, output) = match transaction_outcome {
+			// `with_transactional` executed successfully, and we have the expected output.
+			Ok((success, output)) => (success, output),
+			// `with_transactional` returned an error, and we propagate that error and note no state
+			// has changed.
+			Err(error) => (false, Err(ExecError { error, origin: ErrorOrigin::Callee })),
```
Comment (athei) [inline frame/contracts/src/exec.rs:788]: Should be a `Caller` error.

```suggestion
			Err(error) => (false, Err(error.into()),
```

**Comment by kianenigma** [comment]
sorry I had to force push because I pushed to commits here by mistake, and if I would only revert them, then the diff of any branch on top of this branch would have been messed up.

**Comment by athei** [comment]
This might be totally naive but couldn't we create something like this for use within the runtime where we know hat we are single threaded?

```rust
use core::cell::RefCell;

struct RuntimeMutex<T>(RefCell<T>));

unsafe impl<T> Sync for RuntimeMutex<T>;

impl Deref<T> for RuntimeMutex<T> {
   type Target = RefCell<T>;
   fn deref(&self) -> &Self::Target {
      &self.0
   }
}
```

File: frame/support/src/storage/mod.rs [hunk 2]
```diff
@@ -44,71 +47,88 @@ pub mod types;
...
 }
 
 /// Execute the supplied function in a new storage transaction.
 ///
 /// All changes to storage performed by the supplied function are discarded if the returned
 /// outcome is `TransactionOutcome::Rollback`.
 ///
-/// Transactions can be nested to any depth. Commits happen to the parent transaction.
-pub fn with_transaction<R>(f: impl FnOnce() -> TransactionOutcome<R>) -> R {
+/// Transactions can be nested up to 10 times; more than that will result in an error.
```
Comment (athei) [inline frame/support/src/storage/mod.rs:118]: Isn't `TRANSACTIONAL_LIMIT` set to `255`?

File: frame/support/procedural/src/transactional.rs [hunk 3]
```diff
@@ -49,7 +49,9 @@ pub fn require_transactional(_attr: TokenStream, input: TokenStream) -> Result<T
 	let output = quote! {
 		#(#attrs)*
 		#vis #sig {
-			#crate_::storage::require_transaction();
+			if !#crate_::storage::is_transactional() {
+				return Err(#crate_::pallet_prelude::DispatchError::TransactionalLimit.into());
```
Comment (athei) [inline frame/support/procedural/src/transactional.rs:53]: I think we should have a separate error here for "Layer expected". "Too Many layers" is quite a different error.

**Review by athei — APPROVED** [review]

File: frame/support/src/storage/mod.rs [hunk 4]
```diff
@@ -44,71 +47,123 @@ pub mod types;
...
-	use std::cell::RefCell;
+mod transaction_level_tracker {
+	type Layer = u32;
+	const TRANSACTION_LEVEL_KEY: &'static [u8] = b":transaction_level:";
+	const TRANSACTIONAL_LIMIT: Layer = 255;
 
-	thread_local! {
-		static TRANSACTION_LEVEL: RefCell<u32> = RefCell::new(0);
+	pub fn get_transaction_level() -> Layer {
+		crate::storage::unhashed::get_or_default::<Layer>(TRANSACTION_LEVEL_KEY)
```
Comment (xlc) [inline frame/support/src/storage/mod.rs:56]: can we use a global variable, well, for a non persistent global variable?

Comment (kianenigma) [reply frame/support/src/storage/mod.rs:56]: You mean something like this? https://github.com/paritytech/substrate/compare/master...kiz-shawntabrizi-transactional-rework2

Comment (xlc) [reply frame/support/src/storage/mod.rs:56]: yes

**Review by kianenigma — APPROVED** [review]
LGTM. I explored using `environmental` (akin to `thread_local!` globals) a bit instead of temp storage, but it seemed to imply that we have to wrap the outermost dispatch call, somewhere in `executive`, with a function to bring the global reference into scope, and this seemed like a bad design choice to me. The nice thing about using a temp storage is that it need not expose this detail to executive, and the existence of the transactional layer is only contained in frame-support and storage crates. 

I think using the `environmental` might become more and more sensible over time if we have more global variable that we want to exist in scope inside the runtime. So, instead of `environmental!(transaction_layers: u32)`, we would have `environmental!(temp_storage: BTreeMap<Vec<u8>, Vec<u8>>)`, and this environmental reference is initialized in executive, and can store any temporary data that the runtime needs only within a single invocation. 

Alternatively, I wonder if it is too much of a bad pattern to make the assumption that a FRAME runtime will always be single-threaded, thus we can create a global `static mut` variable and mutate it on the fly. Of course, this will be `unsafe`.

**Comment by xlc** [comment]
I don't really think we are going to have multi threaded execution within a single WASM runtime?
We do have experimental support of multi threaded execution and that's executing multiple WASM runtime concurrently so the assumption will still hold true.

**Comment by athei** [comment]
> I don't really think we are going to have multi threaded execution within a single WASM runtime?

I agree. The only mechanism that exists constitutes a pure function (no shared memory with the spawner). Apart from that the current thinking is that partiality happens on the validator level (validating multiple shards at once) rather than per runtime.

> Alternatively, I wonder if it is too much of a bad pattern to make the assumption that a FRAME runtime will always be single-threaded, thus we can create a global static mut variable and mutate it on the fly. Of course, this will be unsafe.

I think we should have that. Still need to care for re-entry, though. This is why we can't just unsafely access mutable globals but must use `RefCell` and `Cell`. We should newtype it and implement `Sync` on it.

File: frame/support/src/storage/mod.rs (hunk 4 above)
Comment (athei) [reply frame/support/src/storage/mod.rs:56]: I think environmental is very cumbersome in this case. Using storage is clearly not the right way to do it but it works for now. In this case we could just use atomics which compile into plain arithmetic for wasm.

**Comment by athei** [comment]
I pushed a change that uses atomics to implement the counter. They compile to just basic arithmetic in wasm. Native runtime will still use the more expensive synchronization instructions but we shouldn't optimize for this case. The proper solution would probably be having `Sync` wrappers around `Cell` and `RefCell`.

Just revert my commit if you don't like it.

**Comment by shawntabrizi** [comment]
Thank you @athei

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
bot merge

File: frame/support/src/storage/mod.rs [hunk 5]
```diff
@@ -44,71 +47,121 @@ pub mod types;
...
 pub mod weak_bounded_vec;
 
-#[cfg(all(feature = "std", any(test, debug_assertions)))]
-mod debug_helper {
-	use std::cell::RefCell;
+mod transaction_level_tracker {
+	use core::sync::atomic::{AtomicU32, Ordering};
+
+	type Layer = u32;
+	static NUM_LEVELS: AtomicU32 = AtomicU32::new(0);
```
Comment (bkchr) [inline frame/support/src/storage/mod.rs:54]: The native runtime can be executed from multiple threads.... 

Why don't we just have used the environmental crate then?

Comment (athei) [reply frame/support/src/storage/mod.rs:54]: It is hard to do here because there is no easy common function where to place it.

Comment (bkchr) [reply frame/support/src/storage/mod.rs:54]: @kianenigma had a solution that looked reasonable to me.

**Comment by shawntabrizi** [comment]
fixed up in https://github.com/paritytech/substrate/pull/11163#event-6366032711

**Pull request merged**